File: app/lib/models/pandal.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../utils/constants.dart';

/// Firestore collection: `pandals`
/// Matches the application schema (architecture doc section 3 and data/schema/):
/// id, name, lat, lng, zone, area, region, rating, theme, crowd_level,
/// timings, image_url, description, transport, nearest_metro, nearest_railway
class Pandal {
  Pandal({
    required this.id,
    required this.name,
    required this.lat,
    required this.lng,
    required this.zone,
    required this.theme,
    required this.timings,
    required this.imageUrl,
    required this.description,
    this.area,
    this.region,
    this.rating,
    this.crowdLevel,
    this.nearestMetro,
    this.nearestMetroList = const [],
    this.nearestRailway,
    this.nearestRailwayList = const [],
    this.transport = const [],
    this.specialFeatures = const [],
  });

  final String id;
  final String name;
  final double lat;
  final double lng;
  final KolkataZone zone;
  final String? area;
  final String? region;
  final double? rating;
  final String theme;
  final String? crowdLevel;
  final String timings;
  final String imageUrl;
  final String description;
  final String? nearestMetro;
  final List<String> nearestMetroList;
  final String? nearestRailway;
  final List<String> nearestRailwayList;
  final List<String> transport;
  final List<String> specialFeatures;

  factory Pandal.fromFirestore(DocumentSnapshot<Map<String, dynamic>> doc) {
    return Pandal.fromMap(doc.id, doc.data() ?? {});
  }

  factory Pandal.fromMap(String id, Map<String, dynamic> d) {
    return Pandal(
      id: id,
      name: d['name'] as String? ?? '',
      lat: (d['lat'] as num?)?.toDouble() ?? 0,
      lng: (d['lng'] as num?)?.toDouble() ?? 0,
      zone: _parseZone(d['zone'] as String?),
      area: d['area'] as String?,
      region: d['region'] as String?,
      rating: (d['rating'] as num?)?.toDouble(),
      theme: d['theme'] as String? ?? '',
      crowdLevel: d['crowd_level'] as String? ?? d['crowdLevel'] as String?,
      timings: d['timings'] as String? ?? '',
      imageUrl: d['image_url'] as String? ?? d['imageUrl'] as String? ?? '',
      description: d['description'] as String? ?? '',
      nearestMetro: d['nearest_metro'] as String? ?? d['nearestMetro'] as String?,
      nearestMetroList: _parseStringList(d['nearest_metro_list'] ?? d['nearestMetro']),
      nearestRailway: d['nearest_railway'] as String? ?? d['nearestRailway'] as String?,
      nearestRailwayList: _parseStringList(d['nearest_railway_list'] ?? d['nearestRailway']),
      transport: _parseStringList(d['transport']),
      specialFeatures: _parseStringList(d['special_features'] ?? d['specialFeatures']),
    );
  }

  Map<String, dynamic> toFirestore() => {
        'name': name,
        'lat': lat,
        'lng': lng,
        'zone': zone.name,
        'area': area,
        'region': region,
        'rating': rating,
        'theme': theme,
        'crowd_level': crowdLevel,
        'timings': timings,
        'image_url': imageUrl,
        'description': description,
        'nearest_metro': nearestMetro,
        'nearest_metro_list': nearestMetroList,
        'nearest_railway': nearestRailway,
        'nearest_railway_list': nearestRailwayList,
        'transport': transport,
        'special_features': specialFeatures,
      };

  static KolkataZone _parseZone(String? raw) {
    if (raw == null) return KolkataZone.centralKolkata;
    return KolkataZone.values.firstWhere(
      (z) => z.name.toLowerCase() == raw.toLowerCase(),
      orElse: () => KolkataZone.centralKolkata,
    );
  }

  static List<String> _parseStringList(dynamic raw) {
    if (raw is List) {
      return raw.map((e) => e.toString()).toList();
    }
    return const [];
  }
}

File: ios/Uma/Models/Pandal.swift
// Pandal.swift
// Port of app/lib/models/pandal.dart — 15 fields, snake_case JSON keys.

import Foundation
import CoreLocation

struct Pandal: Identifiable, Codable, Hashable, Sendable {
    let id: String
    let name: String
    let lat: Double
    let lng: Double
    let zone: KolkataZone
    let theme: String
    let timings: String
    let imageUrl: String?
    let description: String
    let area: String?
    let region: String?
    let rating: Double?
    let crowdLevel: String?
    let nearestMetro: String?
    let nearestMetroList: [String]
    let nearestRailway: String?
    let nearestRailwayList: [String]
    let transport: [String]
    let specialFeatures: [String]

    var latitude: Double { lat }
    var longitude: Double { lng }
    var coordinate: CLLocationCoordinate2D { .init(latitude: lat, longitude: lng) }
    var entryFee: String { "Free entry" }
    var zoneLabel: String { zone.label }
    var bengaliName: String? { nil }
    var hasHighRating: Bool { (rating ?? 0) >= 4.5 }
    var isCrowded: Bool { crowdLevel?.lowercased() == "high" }

    enum CodingKeys: String, CodingKey {
        case id, name, lat, lng, zone, theme, timings
        case imageUrl       = "image_url"
        case description
        case area, region, rating
        case crowdLevel     = "crowd_level"
        case nearestMetro   = "nearest_metro"
        case nearestMetroList = "nearest_metro_list"
        case nearestRailway = "nearest_railway"
        case nearestRailwayList = "nearest_railway_list"
        case transport
        case specialFeatures = "special_features"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        name = (try? c.decode(String.self, forKey: .name)) ?? ""
        lat = (try? c.decode(Double.self, forKey: .lat)) ?? 0
        lng = (try? c.decode(Double.self, forKey: .lng)) ?? 0
        let zoneRaw = try? c.decode(String.self, forKey: .zone)
        zone = KolkataZone.parse(zoneRaw)
        theme = (try? c.decode(String.self, forKey: .theme)) ?? ""
        timings = (try? c.decode(String.self, forKey: .timings)) ?? ""
        imageUrl = try? c.decode(String.self, forKey: .imageUrl)
        description = (try? c.decode(String.self, forKey: .description)) ?? ""
        area = try? c.decode(String.self, forKey: .area)
        region = try? c.decode(String.self, forKey: .region)
        rating = try? c.decode(Double.self, forKey: .rating)
        crowdLevel = try? c.decode(String.self, forKey: .crowdLevel)
        nearestMetro = try? c.decode(String.self, forKey: .nearestMetro)
        nearestMetroList = (try? c.decode([String].self, forKey: .nearestMetroList)) ?? []
        nearestRailway = try? c.decode(String.self, forKey: .nearestRailway)
        nearestRailwayList = (try? c.decode([String].self, forKey: .nearestRailwayList)) ?? []
        transport = (try? c.decode([String].self, forKey: .transport)) ?? []
        specialFeatures = (try? c.decode([String].self, forKey: .specialFeatures)) ?? []
    }

    init(id: String, name: String, lat: Double, lng: Double, zone: KolkataZone,
         theme: String, timings: String, imageUrl: String?, description: String,
         area: String?, region: String?, rating: Double?, crowdLevel: String?,
         nearestMetro: String?, nearestMetroList: [String], nearestRailway: String?,
         nearestRailwayList: [String], transport: [String], specialFeatures: [String]) {
        self.id = id; self.name = name; self.lat = lat; self.lng = lng
        self.zone = zone; self.theme = theme; self.timings = timings
        self.imageUrl = imageUrl; self.description = description
        self.area = area; self.region = region; self.rating = rating
        self.crowdLevel = crowdLevel
        self.nearestMetro = nearestMetro; self.nearestMetroList = nearestMetroList
        self.nearestRailway = nearestRailway; self.nearestRailwayList = nearestRailwayList
        self.transport = transport; self.specialFeatures = specialFeatures
    }
}
